src/routes_api.py

```diff
@@ -51,4 +57,204 @@ def trigger_scheduled_run(cid, aid, scheduled_run_id):
+                        return util.error(f"Invalid or non-existent student NetID: {student_netid}")
+            try:
+                config = form["config"]
+                if not isinstance(config, dict):
+                    config = json.loads(config)
+                msg = bw_api.set_assignment_config(cid, f"{aid}_{run_id}", config)
+                if msg:
+                    return util.error(f"Failed to upload config to Broadway: {msg}")
+            except json.decoder.JSONDecodeError:
+                return util.error("Failed to decode config JSON")
+
+            # Schedule a new run with scheduler
+            if scheduled_run_id is None:
+                scheduled_run_id = schedule_run(run_time, cid, aid)
+                if scheduled_run_id is None:
+                    return util.error("Failed to schedule run with scheduler")
+            # Or if the run was already scheduled, update the time
+            else:
+                if not update_scheduled_run(scheduled_run_id, run_time):
+                    return util.error("Failed to update scheduled run time with scheduler")
+
+            assert scheduled_run_id is not None
+
+            if not db.add_or_update_scheduled_run(run_id, cid, aid, run_time, due_time, roster, form["name"], scheduled_run_id):
+                return util.error("Failed to save the changes, please try again.")
+            return util.success("Successfully scheduled run.", HTTPStatus.OK)
+        
+        @blueprint.route("/api/<cid>/<aid>/schedule_run", methods=["POST"])
+        @auth.require_course_auth
+        @auth.require_admin_status
+        def api_add_scheduled_run(cid, aid):
+            # generate new id for this scheduled run
+            form = request.json
+            run_id = db.generate_new_id()
+            return add_or_edit_scheduled_run(cid, aid, run_id, form, None)
+
+        @blueprint.route("/api/<cid>/<aid>/schedule_runs", methods=["POST"])
+        @auth.require_course_auth
+        @auth.require_admin_status
+        def api_add_scheduled_runs(cid, aid):
+            form = request.json
+            # generate new id for this scheduled run
+            missing = util.check_missing_fields(form, "runs")
+            if missing:
+                return util.error(f"Missing fields {', '.join(missing)}")
+            # TODO: there's probably a better way to do this
+            print(form["runs"])
+            print(type(form["runs"]))
+            if not isinstance(form["runs"], list):
+                return util.error("runs field must be a list of run configs!")
+            for run_config in form["runs"]:
+                run_id = db.generate_new_id()
+                retval = add_or_edit_scheduled_run(cid, aid, run_id, run_config, None)
+                # TODO: There should be a better distinction between good and bad responses
+                if retval[1] != HTTPStatus.OK:
```

Comment:
If the run fails, can we just copy the code from `staff_delete_scheduled_run` [here](https://github.com/illinois-cs241/broadway-on-demand/blob/8561c6c877468a072be3a30bf24cd4f22df1814e/src/routes_admin.py#L378-L385) to roll back the previously-scheduled runs?

Reply:
I might need some more clarification on what you mean here, are you saying to delete the run from `sched_api` as a precaution, or are you saying that if any run fails to be added, we should remove all scheduled runs from the assignment? I'm not sure I agree with the second approach as much, in all honesty, but I guess I can see why we would want to do that.